### DISCUSSION TITLE
Is it recommended/best practice to call ERR_clear_error before SSL_read SSL_write et al?

### DISCUSSION THREAD

**Discussion by space88man** [openssl/openssl#23025] [Q&A] [answered]
The error queue management is confusing to me: what is the best practice for OpenSSL 3.x?

* Is this problem only relevant to applications that use threads?
* Folklore has it that you should call `ERR_clear_error` before any call to `SSL_read`, `SSL_write` if you are going to use the result `SSL_get_error`  (e.g flushing BIOs in the non-blocking socket examples): https://stackoverflow.com/questions/18179128/how-to-manage-the-error-queue-in-openssl-ssl-get-error-and-err-get-error
* @mattcaswell  has also written this “ Clearing the error queue as part of every IO call seems like a "nuclear" option and over kill.”  here https://github.com/openssl/openssl/pull/17169#issuecomment-1041352792. Is this referring to the same thing?
* The ddd libuv example (https://github.com/openssl/openssl/blob/master/doc/designs/ddd/ddd-06-mem-uv.c non-blocking/epoll-type application) doesn't call `ERR_clear_error` - because it doesn't use threads?

**Comment by t8m** [comment, marked as answer]
Error queues are separate per thread so it really does not depend on whether the caller is multithreaded or not. You should call ERR_clear_error() if you aren't sure whether there are any errors left on the queue from previous OpenSSL calls.

In general there should be no such errors if there were no failures of previous OpenSSL calls but of course there might be bugs. We treat any error being added to the error queue if an OpenSSL function returns success as bugs and fix these.

However any failed OpenSSL call can potentially leave one or multiple errors on the error queue so you need to clear the error queue if an OpenSSL call failed an you want to subsequently call SSL_read/write.

**Comment by hlandau** [comment]
I'd certainly recommend doing so.
